datalad/distribution/publish.py

```diff
@@ -62,13 +63,23 @@ def _publish_dataset(ds, remote, refspec, paths, annex_copy_options):
     # `refspec`
 
     # Plan:
-    # 1. Check if there is anything to push
-    # 2. If so, process push dependencies
-    # 3. fetch and merge annex branch
-    # 4. Push non-annex branch(es)
-    # 5. copy data
+    # 0. Filter out path to the dataset itself
+    # 1. Check if there is anything to push, and if so
+    #    2. process push dependencies
+    #    3. fetch and merge annex branch
+    #    4. push non-annex branch(es)
+    # 5. copy data to the remote if paths are provided or it wants something generally
 
     published, skipped = [], []
+
+    # pre-treat paths.  If path points to entire dataset, and not its root directory
+    # i.e. dataset or dataset/  vs dataset/. - we do not invoke copying all the data
+    # TODO: move into a helper function and test
+    paths = filter(
+        lambda p: p.rstrip(dirsep) != ds.repo.path.rstrip(dirsep),
```

Comment:
Apart from that, I see the intention of the filtering, but I think this is leading to inconsistency of API and is somewhat confusing. Now passed `paths` are to be published but not if they point to a dataset's root dir, but then it's nevertheless published if it ends with `[/].`? At least provide proper documentation in `Publish._params_` please.